### PR TITLE
Implement saving of captures.

### PR DESCRIPTION
**Pull request by ben-clayton**
The GAPIC code is still very rough, but it works.

Fixes: #112

### CODE REVIEW

**Comment by AWoloszyn** [comment]
The server-side looks good to me. I cannot comment on the client, I do not understand how any of that works.

File: gapic/src/main/com/google/gapid/models/Capture.java [hunk 1]
```diff
@@ -129,6 +130,57 @@ protected void onUiThreadError(GapisInitException error) {
...
+      if (canonicalFile.getParentFile() != null) {
+        settings.lastOpenDir = canonicalFile.getParentFile().getAbsolutePath();
+      }
+    } catch (IOException e) {
+      if (file.getParentFile() != null) {
+        settings.lastOpenDir = file.getParentFile().getAbsolutePath();
+      }
+
+      fireError(new GapisInitException(
+          GapisInitException.MESSAGE_TRACE_FILE_LOAD_FAILED + file, "Loading trace failed", e));
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/models/Capture.java]: Wrong message.

File: gapic/src/main/com/google/gapid/models/Capture.java [hunk 2]
```diff
@@ -129,6 +130,57 @@ protected void onUiThreadError(GapisInitException error) {
...
+              }
+              return success(true);
+            } catch (ExecutionException | RpcException | IOException e) {
+              return error(e.toString());
+            }
+          }
+
+          @Override
+          protected void onUiThreadSuccess(Boolean unused) {
+            LOG.warning("Trace saved.");
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/models/Capture.java]: This should be info level.
```LOG.log(INFO, ...)```

File: gapic/src/main/com/google/gapid/models/Capture.java [hunk 3]
```diff
@@ -129,6 +130,57 @@ protected void onUiThreadError(GapisInitException error) {
...
+          }
+
+          @Override
+          protected void onUiThreadError(String error) {
+            LOG.warning("Couldn't save trace: " + error);
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/models/Capture.java]: Use
```LOG.log(INFO, "...")```

Where INFO is a static import of Level.INFO

File: gapic/src/main/com/google/gapid/models/Capture.java [hunk 4]
```diff
@@ -129,6 +130,57 @@ protected void onUiThreadError(GapisInitException error) {
...
+          @Override
+          protected ResultOrError<Boolean, String> onRpcThread(Result<byte[]> result) throws RpcException, ExecutionException {
+            try {
+              byte[] data = result.get();
+              try (FileOutputStream fos = new FileOutputStream(file)) {
+                fos.write(data);
+              }
+              return success(true);
+            } catch (ExecutionException | RpcException | IOException e) {
+              return error(e.toString());
```
Comment (pmuetschard) [inline gapic/src/main/com/google/gapid/models/Capture.java]: Change the error type to be Exception and then log the exception in the error handler.

File: gapic/src/main/com/google/gapid/models/Capture.java (hunk 1 above)
Comment (ben-clayton) [inline gapic/src/main/com/google/gapid/models/Capture.java]: Done

File: gapic/src/main/com/google/gapid/models/Capture.java (hunk 4 above)
Comment (ben-clayton) [inline gapic/src/main/com/google/gapid/models/Capture.java]: Done

File: gapic/src/main/com/google/gapid/models/Capture.java (hunk 2 above)
Comment (ben-clayton) [inline gapic/src/main/com/google/gapid/models/Capture.java]: Done

File: gapic/src/main/com/google/gapid/models/Capture.java (hunk 3 above)
Comment (ben-clayton) [inline gapic/src/main/com/google/gapid/models/Capture.java]: Done

**Comment by ben-clayton** [comment]
Pascal - care to take a look?

**Pull request merged**